src/pip/_internal/repositories/entries.py

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import re
+import sys
+
+from pip._vendor import html5lib, six
+
+# Copied from distlib.compat (to match pip's implementation).
+if sys.version_info < (3, 4):
+    unescape = six.moves.html_parser.HTMLParser().unescape
+else:
+    from html import unescape
+
+EGG_INFO_RE = re.compile(r"([a-z0-9_.]+)-([a-z0-9_.!+-]+)", re.IGNORECASE)
+
+
+def match_egg_info_version(egg_info, package_name, _egg_info_re=EGG_INFO_RE):
+    """Pull the version part out of a string.
+
+    :param egg_info: The string to parse. E.g. foo-2.1
+    :param package_name: The name of the package this belongs to. None to
+        infer the name. Note that this cannot unambiguously parse strings
+        like foo-2-2 which might be foo, 2-2 or foo-2, 2.
+    """
+    match = _egg_info_re.search(egg_info)
+    if not match:
+        raise ValueError(egg_info)
+    if package_name is None:
+        return match.group(0).split("-", 1)[-1]
+    name = match.group(0).lower()
+    # To match the "safe" name that pkg_resources creates:
+    name = name.replace('_', '-')
+    # project name and version must be separated by a dash
+    look_for = package_name.lower() + "-"
+    if name.startswith(look_for):
+        return match.group(0)[len(look_for):]
+    return None
+
+
+def _parse_base_url(document, page_url):
+    """Get the base URL of this document.
+
+    This looks for a ``<base>`` tag in the HTML document. If present, its href
+    attribute denotes the base URL of anchor tags in the document. If there is
+    no such tag (or if it does not have a valid href attribute), the HTML
+    file's URL is used as the base URL.
+
+    :param document: An HTML document representation. The current
+        implementation expects the result of ``html5lib.parse()``.
+    :param page_url: The URL of the HTML document.
+    """
+    bases = [
```

Comment:
It might be more elegant to use a for-else here with a `break` if the condition is met. That way you stop at the first valid base and don't need to do `bases[0]`, etc.

Reply:
Agreed. I copied existing pip code to create this function. It would be a good idea to clean it up a bit.